flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/JobManagerTaskRestore.java

```diff
@@ -58,4 +78,12 @@ public String toString() {
                 + taskStateSnapshot
                 + '}';
     }
+
+    public Set<StateObjectID> getSharedStateObjectIDs() {
+        return unmodifiableSet(sharedStateObjectIDs);
+    }
+
+    void addSharedObjectStateID(StateObjectID sharedStateID) {
+        sharedStateObjectIDs.add(sharedStateID);
```

Comment:
This set is mutable, because it's only possible to compute shared state once it is assigned.

It probably can be avoided by doing an additional pass in `StateAssignementOperation` with some "fake" graph objects or collecting the assingments, but this doesn't seem to worth it.